File: benchmarks/dataset.py
"""Load benchmark dataset from data.json — single source of truth shared with Go."""

import json
from pathlib import Path

_DATA_FILE = Path(__file__).parent / "data.json"

with open(_DATA_FILE) as f:
    _raw = json.load(f)

# Expose as list of tuples: (cached_key, lookup_query, answer, expect_hit, category)
DATASET = [(d["cached"], d["lookup"], d["answer"], d["expect_hit"], d["category"]) for d in _raw]

File: benchmarks/dataset.go
package benchmarks

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
)

// QueryPair represents a cached entry and a lookup query.
// ExpectHit indicates whether the lookup is semantically equivalent to the cached key.
type QueryPair struct {
	Cached    string `json:"cached"`
	Lookup    string `json:"lookup"`
	Answer    string `json:"answer"`
	ExpectHit bool   `json:"expect_hit"`
	Category  string `json:"category"`
}

// Dataset contains 100 realistic LLM query pairs for benchmarking.
// Loaded from data.json — the single source of truth shared with the Python benchmark.
var Dataset = mustLoadDataset()

func mustLoadDataset() []QueryPair {
	// Resolve data.json relative to this source file so it works
	// regardless of the working directory (go test, Docker, etc.).
	_, src, _, _ := runtime.Caller(0)
	path := filepath.Join(filepath.Dir(src), "data.json")

	data, err := os.ReadFile(path)
	if err != nil {
		panic("benchmarks: cannot read data.json: " + err.Error())
	}

	var pairs []QueryPair
	if err := json.Unmarshal(data, &pairs); err != nil {
		panic("benchmarks: cannot parse data.json: " + err.Error())
	}
	return pairs
}
